enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,814 @@
+already created. Each CR will be created in the `openshift-ovn-kubernetes` namespace. The name of the CR will be assigned using a hash
+function (similar to the ComputeHash [here](https://github.com/openshift/kubernetes/blob/master/pkg/controller/controller_utils.go#L1157-L1172))
+prefixed by `dns-`. The input to the hash function will be the DNS name. The `.spec.name` field of the CR will be set to the DNS name along with a trailing period.
+The trailing period is added because DNS servers store DNS names with a trailing period. Thus, to make it consistent, and also for ease of matching,
+the `spec.name` will be set with a trailing period.
+* When an OpenShift cluster administrator deletes an EgressFirewall resource for a namespace containing rule(s) for DNS name(s)
+OR updates an EgressFirewall resource for a namespace and deletes rule(s) containing DNS name(s), the cluster manager will receive the delete/update event.
+* The cluster manager will then check if the same DNS names are also used in the EgressFirewall rules in other namespaces. If the
+DNS names are not used, then the cluster manager will delete the corresponding `DNSNameResolver` CRs.
+
+#### Workflow of `DNSNameResolver` controller
+
+The following pseudocode explains the workflow of the `DNSNameResolver` controller:
+```shell
+watch(DNSNameResolver)
+run_dns_resolution_in_background()
+
+if match(event.type, "create") || match(event.type, "update"):
+
+	if check_dns_name_added(obj.spec.name) == false:
+		add_and_resolve_dns_name(obj.spec.name)
+
+	for each resolvedName in obj.status.resolvedNames:
+		ensure_dns_name_added(resolvedName)
+	
+	if removal_of_ips_required(obj):
+		update_dns_resolver(obj)
+		return
+
+	remaining_time_till_grace_period, is_expiring = grace_period_expires_after(obj)
+
+	if is_expiring:
+		reconcile_event_after(remaining_time_till_grace_period)
+		return
+
+else if match(event.type, "delete"):
+
+	if is_wildcard(obj.spec.name) == true
+		|| (is_wildcard(obj.spec.name) == false && match_wildcard_dns_name(obj.spec.name) == false):
+		delete_dns_name(obj.spec.name)
+
+	if is_wildcard(obj.spec.name) == true:
+		for each resolvedName in obj.status.resolvedNames:
+			if resolvedName.dnsName != obj.spec.name && match_regular_dns_name(resolvedName.dnsName) == false:
+				delete_dns_name(obj.spec.name)
+```
+* The `DNSNameResolver` controller will run dns resolution of DNS names in the background which will refresh the `.status` field of the `DNSNameResolver`
+CRs upon expiration of the validity of the corresponding IP addresses associated to the resolved names. As the CoreDNS pods are configured using the upstream
+nameservers present in the node host's `/etc/resolv.conf` file, the upstream nameservers for different CoreDNS pods may serve different IP addresses (possibly
+with different TTLs) for the same DNS name. Thus, while refreshing the information of a DNS name, the DNS lookup requests are sent to a maximum of 5 randomly
```

Comment:
I am not sure if sending 5 requests for every dns name is reasonable, because most of the DNS names will probably be resolved to the same ip. Maybe we can only do that for dnsNames that already have more than 1 resolved ip in the list?

Reply:
We can do that. The idea behind sending the request to 5 coreDNS pods was to try and get the different sets of IPs, if any, for a DNS name.

Reply:
sure, I see the intent, but I am concerned about the performance impact, so trying to see if we can optimize it

Reply:
Doing one lookup should be fine.  How would you resolve things if you got different ip addresses anyway?

Reply:
The idea came out of a discussion with @Miciah. It is mentioned here: https://github.com/openshift/enhancements/pull/1335#discussion_r1280694158

> Doing one lookup should be fine. How would you resolve things if you got different ip addresses anyway?

The DNS resolution will be done for a DNS name using the minimum TTL of all the IP addresses.